[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139220
* #139208
* #139159
* #139090
* #139084
* #139081
* #138275
* __->__ #138005
* #137918
* #139083
* #139082

Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

swolchok committed:
[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures

Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

malfet reviewed (approved):
Sure, but we should have `half_one` and `half_zero` constants somewhere to avoid doing casts all the time

swolchok commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -284,7 +284,55 @@ INSTANTIATE(int8_t);
+  // https://godbolt.org/z/v936hroYb
+  using c10::detail::fp16_ieee_from_fp32_value;;
+  return (trans == 'T' || trans == 't') && incx == 1 &&
+    alpha.x == fp16_ieee_from_fp32_value(1.0f) &&
+    beta.x == fp16_ieee_from_fp32_value(0.0f);
```
> fp16_ieee_from_fp32_value calls are very slow

above comment demonstrates it's constant-foldable

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to gemv fast path for non-aarch64 architectures"


Following up on previous rev to use fp16_gemv_trans in gemv, not just gemm-used-for-gemv.

Differential Revision: [D64351092](https://our.internmc.facebook.com/intern/diff/D64351092/)

[ghstack-poisoned]
